mailbox/plugin/deleted-messages-vault/src/main/java/org/apache/james/mailbox/vault/SupportedChecks.java

```diff
@@ -0,0 +1,158 @@
+            .build();
+
+        @Override
+        default Class<?> supportedType() {
+            return SUPPORTED_TYPE;
+        }
+    }
+
+    interface StringOperator extends Operator {
+        class Contains implements IntegerOperator {
+
+            @Override
+            public OperatorSign operatorSign() {
+                return CONTAINS;
+            }
+
+            @Override
+            public boolean matches(Query.QueryElementValue valueToCheck, ComparedValue comparedValue) {
+                return ((String) comparedValue.value).contains((String) valueToCheck.value);
+            }
+        }
+
+        Class<?> SUPPORTED_TYPE = String.class;
+
+        StringOperator.Contains STRING_CONTAINS = new StringOperator.Contains();
+
+        ImmutableTable<Class<?>, OperatorSign, Operator> SUPPORTED_OPERATORS = ImmutableTable.<Class<?>, OperatorSign, Operator>builder()
+            .put(SUPPORTED_TYPE, STRING_CONTAINS.operatorSign(), STRING_CONTAINS)
+            .build();
+
+        @Override
+        default Class<?> supportedType() {
+            return SUPPORTED_TYPE;
+        }
+    }
+
+    private static ImmutableTable<Class<?>, OperatorSign, Operator> SUPPORTED_CHECKS = ImmutableTable
+        .<Class<?>, OperatorSign, Operator>builder()
+        .putAll(IntegerOperator.SUPPORTED_OPERATORS)
+        .putAll(StringOperator.SUPPORTED_OPERATORS)
+        .build();
+
+    private static boolean matches(Query.QueryElement element, ComparedValue comparedValue) {
+        Preconditions.checkArgument(SUPPORTED_CHECKS.containsRow(element.value.type()), element.value.type().getName() + " is not supported");
+        Preconditions.checkArgument(SUPPORTED_CHECKS.containsRow(comparedValue.type()), comparedValue.type().getName() + " is not supported");
+
+        if (element.value.type().equals(comparedValue.type()) // E.g, String field and Integer field dont have same type, so ignore
+            && element.value.fieldName.equals(comparedValue.fieldName)) { // E.g, deliveryDate field and deletionDate field don't have same name, so ignore
+            return Optional.ofNullable(SUPPORTED_CHECKS.get(element.value.type(), element.operatorSign))
+                .map(operator -> operator.matches(element.value, comparedValue))
+                .orElseThrow(RuntimeException::new);
+        }
+
+        return true; // no match type or field name, then ignore, that means element is qualified, good to be returned to users
+    }
+
+    private static boolean matches(Query query, ComparedValue comparedValue) {
+        return query.elements.stream()
+            .map(element -> matches(element, comparedValue))
+            .reduce(true, (first, second) -> first & second);
+    }
+
+    public static boolean matches(Query query, ComparedValues comparedValues) {
+        return comparedValues.values().stream()
+            .map(comparedValue -> matches(query, comparedValue))
+            .reduce(true, (first, second) -> first & second);
```

Comment:
& is a bytewise AND operator most probably NOT doing what you think it does.

Please stick to the logical `&&` operator in all cases.

We are doing some logical boolean algebra, not some bit manipulation dark magic.